### ISSUE TITLE
Rendering removes colons and text when more than one colon/line is used (Hugo v0.54)

### ISSUE THREAD

**Issue by infosec-handbook** [alecthomas/chroma#232]
As mentioned in gohugoio/hugo#5683, Hugo 0.54 and previous versions remove multiple colons and omit enclosed text when used in {{highlight}} blocks.

Example:
```
{{< highlight apache "linenos=table,linenostart=1" >}}
[…]
# Modern configuration, tweak to your needs
SSLProtocol +TLSv1.2
SSLCipherSuite ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256
SSLHonorCipherOrder on
SSLCompression off
SSLSessionTickets off
SSLOpenSSLConfCmd Curves secp521r1:secp384r1:prime256v1
SSLOpenSSLConfCmd ECDHParameters secp521r1
[…]
{{< / highlight >}}
```

Final HTML file contains:
```
[…]
# Modern configuration, tweak to your needs
SSLProtocol +TLSv1.2
SSLCipherSuite ECDHE-ECDSA-AES256-GCM-SHA384ECDHE-ECDSA-CHACHA20-POLY1305ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256
SSLHonorCipherOrder on
SSLCompression off
SSLSessionTickets off
SSLOpenSSLConfCmd Curves secp521r1prime256v1
SSLOpenSSLConfCmd ECDHParameters secp521r1
[…]
```

As you can see, the lines SSLCipherSuite and SSLOpenSSLConfCmd contain less colons and some enclosed parameters are missing. This is also true for IPv6 addresses in highlight blocks as they contain several colons.

**Comment by Jos512** [comment]
I cannot replicate this unfortunately. But I have the same Hugo version as you have. Can this perhaps be caused by a Hugo setting? Or did I make a stupid mistake in replicating your Apache code? 

## Input in Markdown file
This is what I put into my Markdown file:

```
Code output with 'fences' (`~`):

~~~apache
SSLProtocol +TLSv1.2
SSLCipherSuite ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256
SSLHonorCipherOrder on
SSLCompression off
SSLSessionTickets off
SSLOpenSSLConfCmd Curves secp521r1:secp384r1:prime256v1
SSLOpenSSLConfCmd ECDHParameters secp521r1
~~~

Code output with `highlight`:

{{< highlight apache "linenos=table,linenostart=1" >}}
SSLProtocol +TLSv1.2
SSLCipherSuite ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256
SSLHonorCipherOrder on
SSLCompression off
SSLSessionTickets off
SSLOpenSSLConfCmd Curves secp521r1:secp384r1:prime256v1
SSLOpenSSLConfCmd ECDHParameters secp521r1
{{< / highlight >}}
```

## Output rendered by Hugo
This is the rendered output I got in Hugo:

~~~
Code output with ‘fences’ (~):

SSLProtocol +TLSv1.2
SSLCipherSuite ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256
SSLHonorCipherOrder on
SSLCompression off
SSLSessionTickets off
SSLOpenSSLConfCmd Curves secp521r1:secp384r1:prime256v1
SSLOpenSSLConfCmd ECDHParameters secp521r1

Code output with highlight:

1 2 3 4 5 6 7 | SSLProtocol +TLSv1.2 SSLCipherSuite ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256 SSLHonorCipherOrder on SSLCompression off SSLSessionTickets off SSLOpenSSLConfCmd Curves secp521r1:secp384r1:prime256v1 SSLOpenSSLConfCmd ECDHParameters secp521r1
-- | --
~~~

The colons (`:`) that appear between the values are kept, from what I can tell.

Copying over that line number table did fail, so it's just a mess to look at. :slightly_smiling_face:  But here's a screenshot:

![2019-03-08_15-58-48](https://user-images.githubusercontent.com/13403160/54036134-19db4100-41bb-11e9-8256-8da01a03d3ce.png)

## HTML rendered by Hugo (/Chroma)
The raw HTML that Hugo makes for me is:

~~~html
<div class="highlight"><pre class="chroma"><code class="language-apache" data-lang="apache"><span class="nb">SSLProtocol</span> +TLSv1.2
<span class="nb">SSLCipherSuite</span> ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256
<span class="nb">SSLHonorCipherOrder</span> <span class="k">on</span>
<span class="nb">SSLCompression</span> <span class="k">off</span>
<span class="nb">SSLSessionTickets</span> <span class="k">off</span>
<span class="nb">SSLOpenSSLConfCmd</span> Curves secp521r1:secp384r1:prime256v1
<span class="nb">SSLOpenSSLConfCmd</span> ECDHParameters secp521r1</code></pre></div>
<p>Code output with <code>highlight</code>:</p>

<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code class="language-apache" data-lang="apache"><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span><span class="lnt">6
</span><span class="lnt">7
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-apache" data-lang="apache"><span class="nb">SSLProtocol</span> +TLSv1.2
<span class="nb">SSLCipherSuite</span> ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256
<span class="nb">SSLHonorCipherOrder</span> <span class="k">on</span>
<span class="nb">SSLCompression</span> <span class="k">off</span>
<span class="nb">SSLSessionTickets</span> <span class="k">off</span>
<span class="nb">SSLOpenSSLConfCmd</span> Curves secp521r1:secp384r1:prime256v1
<span class="nb">SSLOpenSSLConfCmd</span> ECDHParameters secp521r1</code></pre></td></tr></table>
</div>
</div>
~~~

## Comparison of output
If I compare the `SSLCipherSuite` line from the opening post of this issue with what Hugo renders for me, they're the same:

~~~
SSLCipherSuite ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256
SSLCipherSuite ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256
~~~

(First line above is from the opening post, the second is what Hugo returns with the `highlight` function.)


----

Hugo version that I use:

~~~
C:\site>hugo env
Hugo Static Site Generator v0.54.0/extended windows/amd64 BuildDate: unknown
GOOS="windows"
GOARCH="amd64"
GOVERSION="go1.11.5"
~~~

**Comment by vladfaust** [comment]
I have the same issue in my [blog post](https://blog.onyxframework.org/posts/creating-json-apis-with-onyx-part-2#adding-the-create-endpoint). Scroll down a little to see my comment:

> I cannot highlight the line in this block due to a bug in Chroma code highlighting

If I try to highlight the line number 8 in this code:

```crystal
require "onyx/rest"

require "./models/**"
require "./views/**"
require "./actions/**"

Onyx.get "/", Actions::Hello
Onyx.post "/items", Actions::Items::Create # << This
Onyx.render(:json)
Onyx.listen
```

Then the `Actions::Items::Create` turns into `Actions::Create` for some reason, i.e. `Items` token is gone.

**Comment by infosec-handbook** [comment]
@Jos512 
I still get the same error, for instance:

```
{{< highlight apache "linenos=table,linenostart=1" >}}
SSLOpenSSLConfCmd Curves secp521r1:secp256v1:secp384r1
{{< /highlight >}}
```

is rendered as `SSLOpenSSLConfCmd Curves secp521r1secp384r1`.

Hugo version:
```
[user@client]$ hugo env
Hugo Static Site Generator v0.54.0/extended linux/amd64 BuildDate: unknown
GOOS="linux"
GOARCH="amd64"
GOVERSION="go1.11.5"
```

**Comment by alecthomas** [comment]
I can't replicate this locally with Chroma:

```
$ chroma -f html --html-only -l crystal < foo.cr
<pre class="chroma"><span class="k">require</span> <span class="s2">&#34;onyx/rest&#34;</span>

<span class="k">require</span> <span class="s2">&#34;./models/**&#34;</span>
<span class="k">require</span> <span class="s2">&#34;./views/**&#34;</span>
<span class="k">require</span> <span class="s2">&#34;./actions/**&#34;</span>

<span class="n">Onyx</span><span class="o">.</span><span class="n">get</span> <span class="s2">&#34;/&#34;</span><span class="p">,</span> <span class="n">Actions</span><span class="o">::</span><span class="n">Hello</span>
<span class="n">Onyx</span><span class="o">.</span><span class="n">post</span> <span class="s2">&#34;/items&#34;</span><span class="p">,</span> <span class="n">Actions</span><span class="o">::</span><span class="n">Items</span><span class="o">::</span><span class="n">Create</span> <span class="c1"># &lt;&lt; This</span>
<span class="n">Onyx</span><span class="o">.</span><span class="n">render</span><span class="p">(</span><span class="ss">:json</span><span class="p">)</span>
<span class="n">Onyx</span><span class="o">.</span><span class="n">listen</span>
</pre>
```

Here's a screenshot of the rendered HTML:

![image](https://user-images.githubusercontent.com/41767/54052136-bf33fa80-4237-11e9-880c-be12140c3bf0.png)


You can  try this yourself by downloading Chroma.

I do not believe this is a Chroma issue.

**Comment by alecthomas** [comment]
Given both @Jos512 and I are unable to replicate this, you should take this back upstream.

**Comment by vladfaust** [comment]
Okay, thanks and sorry for disturbing 🙂

**Comment by Jos512** [comment]
@vladfaust @infosec-handbook Thanks for commenting.

I went through Vlad's site repository on GitHub to compare the code and settings with mine site, since I didn't have the issue.

I can replicate it on Vlad's site when I enable/disable the `enableEmoji` configuration setting. When that option is one, things like `::Items::` are dropped from the output that Hugo's `highlight` shortcode makes. 

It seems that those pieces of code are accidentally processed with Hugo's emoji feature.

So I could replicate it now, but it's not a Chroma issue but Hugo problem. (I updated the issue in the Hugo repository with my findings.)
